### ISSUE TITLE
Sending large quantities of messages

### ISSUE THREAD

**Issue by Dadibom** [amqp/rhea#332]
I want to send a large quantity of messages (100k>1mil) messages to an activemq queue using rhea.

Currently I am creating a new sender, and on "sendable" i send messages until sender.sendable() === false. This only lets me send ~300 messages per second which is obviously to slow to be able to queue a million messages in a reasonable timeframe.

I am looking for a faster way to send these messages. I have tried digging into the azure events hub code (as referenced here: https://github.com/amqp/rhea/issues/81 ) to implement batching but i have been unable to get it to work. 
ActiveMQ tells me these messages have JMS_AMQP_MESSAGE_FORMAT = -2147404032. I don't know if activemq simply doesn't support batches or if my format value is overflowing in the node application.

This is the code I have tried (unsuccessfully):

```
function batchQueue (queue, payloads, batchSize = 10) {
  if (! payloads || payloads.length === 0) {
    console.log("Called batchQueue without payloads");
    return;
  }

  const sender = connection.open_sender('queue://' + queue);

  let i = 0;

  sender.on('sendable', context => {
    const messages = [];
    for (j = 0; i < payloads.length && j < batchSize && sender.sendable(); ++i && ++j) {
      messages.push({
        body: message.data_section(Buffer.from(JSON.stringify(payloads[i]), "utf8")),
      });
    }

    sender.send(message.encode({
      body: message.data_sections(messages.map(message.encode)),
    }), null, 0x80013700);

    if (i >= payloads.length) {
      sender.close();
    }
  });
}
```

Any help would be appreciated. Ideally I'd be able to send thousands of messages per second.

**Comment by grs** [comment]
I don't think ActiveMQ supports batching over AMQP, and certainly not the format as defined for Azure. However thousands of messages per second should be possible without batching. Are the messages persistent? The on_sendable event should be raised whenever the broker issues credit. How frequently are you seeing that event and how many messages are you able to send each time?

**Comment by Dadibom** [comment]
Yes, they are persistent.
The sendable event is triggered approx every 2 sec and lets me send ~700 messages every time.

Unrelated but I also noticed the messages are stored in binary, i would like to store the messages in text because I got stomp consumers that just want to parse the message (json) directly. Any way to do this with rhea?

```
warn: 2048 messages sent
warn: sendable event
warn: 700 messages sent
warn: sendable event
warn: 700 messages sent
warn: sendable event
warn: 700 messages sent
warn: sendable event
warn: 700 messages sent
warn: sendable event
warn: 700 messages sent
```

**Comment by grs** [comment]
Set the body to the result of JSON.stringfiy(obj) and it should be encoded as a string.

Regarding the throughput, I would advise asking on the activemq users list. 300 msgs/sec is very slow even for durable, unless you have an extremely slow disk or the messages are all huge. I don't know enough about activemq to offer much advice on how to further debug that though.

**Comment by Dadibom** [comment]
Oops, i forgot json.stringify after reverting the batch test. however i am still getting a binary prefix: SpESsESw�{"val":103}
These messages don't need to be durable if that is faster. I just care about speed for this scenario. How can I accomplish that? As you can see the messages are very small.

I will look for an activemq group too, thanks.

**Comment by grs** [comment]
Non-durable should certainly be faster. Messages should be non-durable by default I believe so if you are not setting durable on the the message to true (or set it to false explicitly) then they should not be written to disk.

The message as you see it is in AMQP format. You can have the broker convert that using a transformer: https://activemq.apache.org/amqp, I think you would want the jms transformer.

**Comment by Dadibom** [comment]
Unfortunately I cannot seem to make them non-persistent.

Tried this but with no change:
```
const sender = connection.open_sender({
    target: {
      address: 'queue://' + queue,
      durable: false,
    },
  });
```

**Comment by Dadibom** [comment]
I now set it on the message instead of the sender and it works, I am now reaching ~50k messages per second!

**Comment by gtully** [comment]
peeking at Artemis, the flow handling is based on a threshold of 300 by default, 1k initial credit, then 700 after that once that many are settled. Sending pre-settled should ensure that credit is replenished exactly on the threshold.

https://github.com/apache/activemq-artemis/blob/master/artemis-protocols/artemis-amqp-protocol/src/main/java/org/apache/activemq/artemis/protocol/amqp/proton/ProtonAbstractReceiver.java#L166

Those values can be configured on the acceptor:
<acceptor name="amqp">tcp://0.0.0.0:5672?amqpIdleTimeout=0;tcpSendBufferSize=1048576;tcpReceiveBufferSize=1048576;protocols=AMQP;useEpoll=true;amqpCredits=1000;amqpMinCredits=300;directDeliver=false;batchDelay=10</acceptor>

**Comment by Dadibom** [comment]
I'm on ActiveMQ 5, but perhaps that's unchanged in artemis. What do you mean by sending "pre-settled"?

**Comment by gtully** [comment]
pre settle is an AMQP link option, in rhea: snd_settle_mode, the send message does expect server to reply to the send, the send is fire and forget.
It corresponds to a delivery expectation of at most once, fast but nothing guaranteed.
There is an example of usage in the tests:
https://github.com/amqp/rhea/blob/master/test/links.ts#L842 and a good discussion in this issue: https://github.com/amqp/rhea/issues/72#issuecomment-393977877

**Comment by ulisesbocchio** [comment]
@Dadibom when you say "I now set it on the message instead of the sender and it works" what do you mean? setting `durable: false` on the message envelope? I have the opposite problem. I see messages as non-persistent and I want them to be persistent....
